### PR TITLE
Fix PayPal UTF characters in product names

### LINKED ISSUE #21463: UTF characters in product names are not correctly passed to PayPal
**Describe the bug**
When I set product name in Russian, it is displayed correctly on the product page, in WooCommerce order line items, etc. However, both PayPal logs and PayPal dashboard show some UTF gibberish instead.

**To Reproduce**
1. Create a product called `Шампунь для увлажнения и питания волос`
2. Enable logging in PayPal
3. Purchase the product through PayPal sandbox
4. Check PayPal dashboard and logs and observe how the product name appears in there.

**Screenshots**

Order page in WooCommerce:

Screenshot
Screenshot: http://cld.wthms.co/dx9CEY

PayPal logs in WooCommerce:

Screenshot
Screenshot: http://cld.wthms.co/04DVQO

PayPal dashboard:

Screenshot
Screenshot: http://cld.wthms.co/08GoXM

**Expected behavior**
The product name is passed correctly to PayPal, even with international characters being used. Looks like `rawurlencode` followed by `rawurldecode` changes the original text in Russian into `–®–∞–º–ø—É–Ω—å –¥–ª—è —É–≤–ª–∞–∂–Ω–µ–Ω–∏—è %`, but I haven't found yet what we can do to fix it.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>

```
### WordPress Environment ###

Home URL: http://wootest.local
Site URL: http://wootest.local
WC Version: 3.5.0
Log Directory Writable: ✔
WP Version: 4.9.8
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.10.1
PHP Version: 7.0.3 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 1,000 MB
PHP Time Limit: 900
PHP Max Input Vars: 3000
cURL Version: 7.38.0
OpenSSL/1.0.1t

SUHOSIN Installed: –
MySQL Version: ❌ 5.5.55-0+deb8u1 - We recommend a minimum MySQL version of 5.6. See: WordPress requirements
Max Upload Size: 1,000 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.5.0
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 26.81MB
Database Data Size: 15.43MB
Database Index Size: 11.38MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.09MB + Index: 0.09MB
wp_woocommerce_tax_rates: Data: 7.52MB + Index: 7.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.08MB + Index: 0.09MB
wp_followup_coupons: Data: 0.02MB + Index: 0.03MB
wp_followup_coupon_logs: Data: 0.02MB + Index: 0.05MB
wp_followup_customers: Data: 0.02MB + Index: 0.06MB
wp_followup_customer_carts: Data: 0.02MB + Index: 0.03MB
wp_followup_customer_notes: Data: 0.02MB + Index: 0.03MB
wp_followup_customer_orders: Data: 0.02MB + Index: 0.02MB
wp_followup_email_coupons: Data: 0.02MB + Index: 0.02MB
wp_followup_email_excludes: Data: 0.02MB + Index: 0.05MB
wp_followup_email_logs: Data: 0.02MB + Index: 0.05MB
wp_followup_email_orders: Data: 1.52MB + Index: 0.47MB
wp_followup_email_order_coupons: Data: 0.02MB + Index: 0.02MB
wp_followup_email_tracking: Data: 0.02MB + Index: 0.08MB
wp_followup_followup_history: Data: 0.02MB + Index: 0.05MB
wp_followup_order_categories: Data: 0.02MB + Index: 0.03MB
wp_followup_order_items: Data: 0.02MB + Index: 0.05MB
wp_followup_subscribers: Data: 0.02MB + Index: 0.03MB
wp_followup_subscribers_to_lists: Data: 0.02MB + Index: 0.03MB
wp_followup_subscriber_lists: Data: 0.02MB + Index: 0.03MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 2.02MB + Index: 0.11MB
wp_postmeta: Data: 2.52MB + Index: 2.02MB
wp_posts: Data: 0.33MB + Index: 0.13MB
wp_snippets: Data: 0.02MB + Index: 0.00MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.09MB + Index: 0.08MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.05MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wcpv_commissions: Data: 0.02MB + Index: 0.00MB
wp_wcpv_per_product_shipping_rules: Data: 0.02MB + Index: 0.00MB
wp_wc_booking_relationships: Data: 0.02MB + Index: 0.03MB
wp_wc_deposits_payment_plans: Data: 0.02MB + Index: 0.00MB
wp_wc_deposits_payment_plans_schedule: Data: 0.02MB + Index: 0.02MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_points_rewards_user_points: Data: 0.02MB + Index: 0.03MB
wp_wc_points_rewards_user_points_log: Data: 0.02MB + Index: 0.05MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_wpml_mails: Data: 0.31MB + Index: 0.00MB

### Post Type Counts ###

attachment: 85
course: 1
follow_up_email: 1
global_product_addon: 1
lesson: 5
page: 10
polylang_mo: 2
post: 3
product: 125
product_variation: 115
question: 3
quiz: 5
revision: 8
scheduled-action: 24
shop_coupon: 3
shop_order: 66
shop_order_refund: 1
shop_subscription: 2
wc_booking: 12
wc_membership_plan: 1
wc_order_status: 7
wc_user_membership: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.5.0-beta.1

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: accommodation-booking (accommodation-booking)
booking (booking)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #9 - /shop/
Cart: #10 - /cart/
Checkout: #11 - /checkout/
My account: #12 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Seventeen
Version: 1.7
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
```

</details>

### PR DESCRIPTION
**Pull request by mikejolley**
For whatever reason (I couldn't work it out exactly) url decoding and encoding breaks UTF-8 characters. When we truncate strings to send to PayPal, this is performed causing invalid display on the PayPal side.

We can avoid the urldecode/encode by limiting strings with mb_ functions which are character encoding aware. This PR implements that.

Fixes #21463

To test, create a product with a long name containing unicode chars e.g. `Шампунь для увлажнения и питания волос Шампунь для увлажнения и питания волос`

On PayPal, after this patch, you'll see a correct truncated version.

String will appear less than 127 in length because of the mb_ functions. This is expected behavior,

> Fix PayPal item name encoding.

### CODE REVIEW

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php [hunk 1]
```diff
@@ -91,12 +91,15 @@ public function get_request_url( $order, $sandbox = false ) {
...
-			$string = rawurldecode( substr( $url_encoded_str, 0, $limit - 3 ) . '...' );
+		$str_limit = $limit - 3;
+		if ( function_exists( 'mb_strimwidth' ) ) {
+			if ( mb_strlen( $string ) > $limit ) {
+				$string = mb_strimwidth( $string, 0, $str_limit ) . '...';
+			}
+		} else {
+			if ( strlen( $string ) > $limit ) {
+				$string = substr( $string, 0, $str_limit ) . '...';
+			}
```
Comment (peterfabian) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php]: This `rawurlencode` + `rawurldecode` were added to make sure that urlencoded string is shorter than 127 characters (based on https://developer.paypal.com/docs/classic/paypal-payments-standard/integration-guide/Appx_websitestandard_htmlvariables/?mark=night_phone_b#individual-items-variables). When looking at it now, I'm not sure if that was a correct reading of the spec.

It seems though, that PayPal does not care about the length that much anyway, so I'm not sure if it's an actual problem or not (the example string ends up as 410 characters when URL-encoded). The code previously cut off the string at byte 124 which could produce an invalid UTF8 character at the end of the string.

Side note: Funnily enough, it seems that item_name is processed by `wc_trim_string` first, which have an interesting side effect: it first checks if number of bytes in the string is larger than the limit (which it is, as the 77 UTF8 chars in the example are not from Latin alphabet), but then trims the string using multibyte char function, so instead of trimming the string, it just appends `...` at the end, so `Шампунь для увлажнения и питания волос Шампунь для увлажнения и питания волос` becomes `Шампунь для увлажнения и питания волос Шампунь для увлажнения и питания волос...`.

Long story short, this seems to work fine, not sure if we need to somehow limit the number of characters in other way, if that's what the PP docs says.

**Review by peterfabian — COMMENTED** [review]
Just a small question, otherwise it seems to work fine.
Plus it needs an update from master.

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php (hunk 1 above)
Comment (mikejolley) [reply includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php]: I'm not sure they care either. In testing I went over the limit without it erring. I do think this keeps it down to the character limit AFTER decoding entities though so I think it's good enough...

Do you think the ... appending thing is something that needs to be looked at further before this PR is approved?

Comment (peterfabian) [reply includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php]: No, that's just an unimportant detail, let's get this in.

**Review by peterfabian — APPROVED** [review]
👍
